File: scripts/analyze-themes.mjs
#!/usr/bin/env node
import { readFileSync, readdirSync } from "fs";
import { join } from "path";

const dir = new URL("../themes", import.meta.url).pathname;
const files = readdirSync(dir).filter(f => f.endsWith(".json") && f !== "dusk-hc.json").sort();

const h2r = h => { h = h.replace("#",""); if(h.length>6) h=h.slice(0,6); return [parseInt(h.slice(0,2),16),parseInt(h.slice(2,4),16),parseInt(h.slice(4,6),16)]; };
const L = rgb => 0.2126*((rgb[0]/255)<=0.03928?rgb[0]/255/12.92:Math.pow((rgb[0]/255+0.055)/1.055,2.4)) + 0.7152*((rgb[1]/255)<=0.03928?rgb[1]/255/12.92:Math.pow((rgb[1]/255+0.055)/1.055,2.4)) + 0.0722*((rgb[2]/255)<=0.03928?rgb[2]/255/12.92:Math.pow((rgb[2]/255+0.055)/1.055,2.4));
const CR = (a,b) => { const l1=L(h2r(a)),l2=L(h2r(b)); return Math.round(((Math.max(l1,l2)+0.05)/(Math.min(l1,l2)+0.05))*100)/100; };
const S = h => h.length > 7 ? h.slice(0,7) : h;

const basic = ["class","function","variable","keyword","string","comment","number","type","interface","namespace"];
const results = [];

for (const f of files) {
  const j = JSON.parse(readFileSync(join(dir, f), "utf8"));
  const c = j.colors || {}, s = j.semanticTokenColors || {}, t = j.tokenColors || [];
  const nm = j.name || f;
  const tp = j.type || (L(h2r(S(c["editor.background"]||"#000"))) > 0.4 ? "light" : "dark");
  const eb = S(c["editor.background"]||"#000"), ef = S(c["editor.foreground"]||"#fff");
  const sb = S(c["sideBar.background"]||""), pb = S(c["panel.background"]||"");
  const sug = c["editorSuggestWidget.background"] ? S(c["editorSuggestWidget.background"]) : "";
  const hov = c["editorHoverWidget.background"] ? S(c["editorHoverWidget.background"]) : "";
  const cr = CR(ef, eb);
  const sbD = sb ? Math.abs(L(h2r(eb)) - L(h2r(sb))) : 0;
  const pbD = pb ? Math.abs(L(h2r(eb)) - L(h2r(pb))) : 0;
  const semK = Object.keys(s);
  const hasSem = basic.every(k => semK.includes(k));
  const missTT = !sug || !hov;
  const is = [], st = [];

  if (cr >= 7) st.push("AAA(" + cr + ")");
  else if (cr >= 4.5) st.push("AA(" + cr + ")");
  else is.push("LOW(" + cr + ")");

  if (sbD < 0.005) st.push("sb=ed");
  else if (sbD > 0.02) is.push("sbGap(" + Math.round(sbD*1000)/10 + "%)");

  if (!missTT) st.push("ttSet");
  else is.push("noTT");

  if (hasSem) st.push("fullSem");
  else is.push("partSem");

  if (t.length >= 25) st.push("richTM(" + t.length + ")");
  else if (t.length < 15) is.push("sparseTM(" + t.length + ")");

  if (pbD < 0.015) st.push("panelOk");
  else if (pbD > 0.03) is.push("panelGap(" + Math.round(pbD*1000)/10 + "%)");

  results.push({ nm, tp, cr, is, st, sc: semK.length, tc: t.length });
}

results.sort((a,b) => a.is.length - b.is.length || b.st.length - a.st.length);

console.log("DEEP ANALYSIS — " + files.length + " THEMES\n");
for (const r of results) {
  const ic = r.is.length === 0 ? "✅" : r.is.length <= 2 ? "⚠️" : "❌";
  console.log(ic + " " + r.nm + " [" + r.tp + "] CR:" + r.cr + " sem:" + r.sc + " tm:" + r.tc);
  if (r.st.length) console.log("  + " + r.st.join(" | "));
  if (r.is.length) console.log("  - " + r.is.join(" | "));
}

const p = results.filter(r => r.is.length === 0).length;
const w = results.filter(r => r.is.length > 0 && r.is.length <= 2).length;
const b = results.filter(r => r.is.length > 2).length;
console.log("\nSUMMARY: ✅" + p + " ⚠️" + w + " ❌" + b);

File: python/dusk_office/analyze_themes.py
#!/usr/bin/env python3
"""Deep contrast, surface, and token coverage analysis for Dusk Office themes."""

from __future__ import annotations

import argparse
import json
from pathlib import Path

from dusk_office.color import contrast_ratio_hex, luminance_from_hex, solid_hex

BASIC_SEMANTIC = [
    "class",
    "function",
    "variable",
    "keyword",
    "string",
    "comment",
    "number",
    "type",
    "interface",
    "namespace",
]

SKIP_FILES = {"dusk-hc.json"}


def repo_root() -> Path:
    return Path(__file__).resolve().parents[2]


def analyze_theme(path: Path, data: dict) -> dict:
    colors = data.get("colors") or {}
    semantic = data.get("semanticTokenColors") or {}
    token_colors = data.get("tokenColors") or []

    name = data.get("name") or path.name
    editor_bg = solid_hex(colors.get("editor.background", "#000"))
    editor_fg = solid_hex(colors.get("editor.foreground", "#fff"))
    sidebar_bg = solid_hex(colors.get("sideBar.background", ""))
    panel_bg = solid_hex(colors.get("panel.background", ""))
    suggest_bg = colors.get("editorSuggestWidget.background")
    hover_bg = colors.get("editorHoverWidget.background")

    theme_type = data.get("type")
    if not theme_type:
        theme_type = "light" if luminance_from_hex(editor_bg) > 0.4 else "dark"

    cr = contrast_ratio_hex(editor_fg, editor_bg)
    sidebar_delta = abs(luminance_from_hex(editor_bg) - luminance_from_hex(sidebar_bg)) if sidebar_bg else 0.0
    panel_delta = abs(luminance_from_hex(editor_bg) - luminance_from_hex(panel_bg)) if panel_bg else 0.0

    semantic_keys = set(semantic.keys())
    has_full_semantic = all(key in semantic_keys for key in BASIC_SEMANTIC)
    missing_tooltips = not suggest_bg or not hover_bg

    strengths: list[str] = []
    issues: list[str] = []

    if cr >= 7:
        strengths.append(f"AAA({cr})")
    elif cr >= 4.5:
        strengths.append(f"AA({cr})")
    else:
        issues.append(f"LOW({cr})")

    if sidebar_delta < 0.005:
        strengths.append("sb=ed")
    elif sidebar_delta > 0.02:
        issues.append(f"sbGap({round(sidebar_delta * 1000) / 10}%)")

    if not missing_tooltips:
        strengths.append("ttSet")
    else:
        issues.append("noTT")

    if has_full_semantic:
        strengths.append("fullSem")
    else:
        issues.append("partSem")

    token_count = len(token_colors)
    if token_count >= 25:
        strengths.append(f"richTM({token_count})")
    elif token_count < 15:
        issues.append(f"sparseTM({token_count})")

    if panel_delta < 0.015:
        strengths.append("panelOk")
    elif panel_delta > 0.03:
        issues.append(f"panelGap({round(panel_delta * 1000) / 10}%)")

    return {
        "name": name,
        "type": theme_type,
        "contrast": cr,
        "issues": issues,
        "strengths": strengths,
        "semantic_count": len(semantic_keys),
        "token_count": token_count,
    }


def run_analysis(themes_dir: Path) -> list[dict]:
    files = sorted(
        path
        for path in themes_dir.glob("*.json")
        if path.name not in SKIP_FILES
    )
    results = []
    for path in files:
        data = json.loads(path.read_text(encoding="utf-8"))
        results.append(analyze_theme(path, data))
    results.sort(key=lambda item: (len(item["issues"]), -len(item["strengths"])))
    return results


def status_icon(issue_count: int) -> str:
    if issue_count == 0:
        return "OK"
    if issue_count <= 2:
        return "WARN"
    return "BAD"


def print_report(results: list[dict]) -> None:
    print(f"DEEP ANALYSIS - {len(results)} THEMES\n")
    for item in results:
        issue_count = len(item["issues"])
        icon = status_icon(issue_count)
        print(
            f"{icon} {item['name']} [{item['type']}] "
            f"CR:{item['contrast']} sem:{item['semantic_count']} tm:{item['token_count']}"
        )
        if item["strengths"]:
            print("  + " + " | ".join(item["strengths"]))
        if item["issues"]:
            print("  - " + " | ".join(item["issues"]))

    perfect = sum(1 for item in results if not item["issues"])
    warning = sum(1 for item in results if 0 < len(item["issues"]) <= 2)
    bad = sum(1 for item in results if len(item["issues"]) > 2)
    print(f"\nSUMMARY: OK={perfect} WARN={warning} BAD={bad}")


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description="Analyze Dusk Office theme quality metrics.")
    parser.add_argument(
        "--themes-dir",
        type=Path,
        default=repo_root() / "themes",
        help="Directory containing theme JSON files",
    )
    args = parser.parse_args(argv)
    if not args.themes_dir.is_dir():
        parser.error(f"themes directory not found: {args.themes_dir}")
    print_report(run_analysis(args.themes_dir))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
